src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md

```diff
@@ -0,0 +1,345 @@
+![](images/Full_Cluster_Msgr_Thread_Scaling_-_FIO_4KB_IOPS_EC62.svg)
+
+We could achieve well over 500GiB/s for reads and nearly 400GiB/s for writes with 4-5 async msgr threads.  But why are the read results so much slower with EC than with replication?  With replication, The primary OSD for a PG only has to read local data and send it to the client.  The network overhead is essentially 1X.  With 6+2 erasure coding, The primary must read 7 or of the 8 chunks from replicas before it can then send the constructed object to the client.  The overall network overhead for the request is roughly (1 + 7/8)X.  That's why we see slightly better than half the performance of 3X replication for reads.  We have the opposite situation for writes.  With 3X replication, the client sends the object to the primary, which then further sends copies over the network to two secondaries. This results in an aggregate network overhead of 3X.  In the EC case, we only need to send 7/8 chunks to the secondaries just like in the read case.  For large writes, performance is actually faster. 
+
+IOPS however, are another story.  For very small reads and writes, Ceph will contact all participating OSDs in a PG for that object even when the data they store isn't relevant for the operation.  For instance if you are doing 4K reads and the data you are interested in is entirely stored in a single chunk on one of the OSDs, Ceph will still fetch data from all OSDs participating in the stripe.  In the summer of 2023, Clyso [resurrected](https://github.com/ceph/ceph/pull/52746) a PR from Xiaofei Cui that implements partial stripe reads for erasure coding to avoid this extra work.  The effect is dramatic:
+
+![](images/FIO_4KB_Random_Read_IOPS_16_Client.svg)
+![](images/FIO_4KB_Random_Read_Cycles_OP_16_Client.svg)
+
+It's not clear yet if we will be able to get this merged for Squid, though Radoslaw Zarzynski, core lead for the Ceph project, has offered to help try to get this over the finish line.
+
+# Squeezing in Msgr Encryption Testing
+
+Finally, we wanted to provide the customer with a rough idea how much msgr level encryption would impact their cluster if they decided to use it.  The adrenaline of the previous night had long faded and I was dead tired at this point.  I managed to run through both 3X replication and 6+2 erasure coding tests with msgr v2 encryption enabled and compared it against our previous test results.
+
+![](images/Full_Cluster_Msgr_Encryption_-_FIO_4MB_Throughput.svg)
+![](images/Full_Cluster_Msgr_Encryption_-_FIO_4KB_IOPS.svg)
+
+The biggest hit is to large reads.  They drop from ~1 TiB/s to around 750 GiB/s.  Everything else sees a more modest, though consistent hit.  At this point I had to stop.  I really wanted to do PG scaling tests and even kernel RBD tests, but it was time to hand the systems back to the customer for re-imaging, followed by one of my excellent colleagues at Clyso to begin the integration process.
```

Comment:
at Clyso beginning the integration

Reply:
Didn't like how that sounded.  How about this?

"...but it was time to hand the systems back to the customer for re-imaging. Once completed, one of my excellent colleagues at Clyso would begin the integration process."

Reply:
What about "for re-imaging, then to one of my excellent colleagues at Clyso for integration" ?

Reply:
I can role with that